File: scripts/traditional_compare.py
"""Kuwana — "Traditional comparison" engine.

A plain, rules-based comparator with **no AI / LLM in the loop**. It mirrors
the transparent heuristics the notebooks use (`recommendation_engine.py` /
`value_score_baseline.py`): a per-category weighted value score, then small
bonuses for special (trending-down / below-median), freshness and verified
providers. Everything is printed in plain text so the app can show exactly
what the engine "thinks" — no black box.

Reads one JSON object from stdin:

    {
      "category_name": "Data bundles",
      "category_slug": "data-bundles",
      "attribute_schema": [
        {"key": "data_amount", "label": "Data", "data_type": "number",
         "unit": "GB", "is_comparable": true, "sort_order": 1}
      ],
      "listings": [
        {
          "id": "...", "name": "...", "provider": "...",
          "provider_verified": true, "price": 5.0, "currency": "USD",
          "attributes": {"data_amount": 5, "validity_days": 30},
          "freshness_status": "fresh",
          "trend_percent": -12.0
        }
      ]
    }

Prints one JSON object to stdout: `{"text": "<human-readable dump>",
"winner": {"name", "score", "why"} | null, "runner_ups": [{"name", "score"}],
"note": "<disclaimer>" | null}`. `text` keeps the full human-readable
breakdown (audit trail / debugging); the other fields are what the app
renders as structured UI instead of dumping `text` verbatim. Uses only the
standard library so it runs anywhere Python 3.8+ is installed — no
pandas/sklearn needed at request time.

Run directly to smoke-test:

    python scripts/traditional_compare.py < sample.json
"""

import json
import sys

# Per-category weights, lifted from notebooks/recommendation_engine.py so the
# on-demand engine agrees with the offline recommender. Categories without an
# entry fall back to the legacy 50/50 price + first-numeric-benefit blend.
CATEGORY_WEIGHTS: dict[str, dict[str, float]] = {
    "data-bundles": {"price": 0.4, "data_amount": 0.4, "validity_days": 0.2},
    "voice-sms-bundles": {"price": 0.35, "minutes": 0.35, "sms_count": 0.2, "validity_days": 0.1},
    "savings-accounts": {"price": 0.3, "interest_rate": 0.4, "monthly_fee": 0.3},
    "current-accounts": {"price": 0.4, "transaction_fee": 0.3, "branch_count": 0.3},
    "motor-insurance": {"price": 0.35, "coverage_amount": 0.45, "claim_turnaround_days": 0.2},
    "life-insurance": {"price": 0.3, "coverage_amount": 0.4, "payout_speed_days": 0.3},
}

FRESHNESS_BONUS = {"fresh": 5, "stale": -6, "unverified": -12}
SPECIAL_BONUS = 20
TRUST_BONUS = 5
TREND_DROP_THRESHOLD = -5.0
BELOW_MEDIAN_FACTOR = 0.9

BAR = "=" * 60


def speak(reasons: list[str]) -> str:
    """Joins reason phrases into one plain sentence: "A", "A and B", or
    "A, B, and C" — read-aloud-friendly instead of a semicolon-delimited log."""
    if not reasons:
        return ""
    if len(reasons) == 1:
        return reasons[0]
    if len(reasons) == 2:
        return f"{reasons[0]} and {reasons[1]}"
    return ", ".join(reasons[:-1]) + f", and {reasons[-1]}"


def _num(value):
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def normalize(values: list[float], invert: bool) -> list[float]:
    lo, hi = min(values), max(values)
    if lo == hi:
        return [100.0 for _ in values]
    return [round((1 - (v - lo) / (hi - lo)) * 100, 1) if invert else round(((v - lo) / (hi - lo)) * 100, 1) for v in values]


def first_numeric_benefit_key(schema: list[dict]) -> str | None:
    numeric = [a for a in schema if a.get("data_type") == "number" and a.get("is_comparable") and a.get("key") != "price"]
    numeric = sorted(numeric, key=lambda a: a.get("sort_order", 0))
    return numeric[0]["key"] if numeric else None


def fmt_attr(listing: dict, key: str, schema: list[dict]) -> str:
    value = listing.get("attributes", {}).get(key)
    if value is None:
        return "—"
    field = next((a for a in schema if a["key"] == key), None)
    unit = field.get("unit") if field else None
    return f"{value} {unit}".strip() if unit else f"{value}"


def compute_value_score(
    listing: dict,
    category_slug: str,
    schema: list[dict],
    prices: list[float],
    compared_set: list[dict],
) -> tuple[float, str]:
    weights = CATEGORY_WEIGHTS.get(category_slug)
    attrs = listing.get("attributes", {})
    price = _num(listing.get("price"))
    if price is None:
        return 0.0, "no price"

    price_scores = normalize(prices, invert=True)
    price_score = price_scores[prices.index(price)] if price in prices else 100.0

    def _column(key: str) -> list[float]:
        return [v for v in (_num(l.get("attributes", {}).get(key)) for l in compared_set) if v is not None]

    def _score_against_set(key: str) -> float:
        col = _column(key)
        raw = _num(attrs.get(key))
        if raw is None or len(col) < 2:
            return 100.0
        norm = normalize(col, invert=False)
        return norm[col.index(raw)] if raw in col else 100.0

    if weights:
        parts: list[float] = []
        breakdown: list[str] = []
        for key, weight in weights.items():
            if key == "price":
                parts.append(price_score * weight)
                breakdown.append(f"price {weight:.0%}")
                continue
            raw = _num(attrs.get(key))
            if raw is None:
                continue
            parts.append(_score_against_set(key) * weight)
            breakdown.append(f"{key} {weight:.0%}")
        if parts:
            score = round(sum(parts), 1)
            return score, " · ".join(breakdown)
        return price_score, "price only"

    benefit_key = first_numeric_benefit_key(schema)
    if benefit_key is None:
        return price_score, "price only (no benefit attribute)"
    raw = _num(attrs.get(benefit_key))
    if raw is None:
        return price_score, "price only"
    benefit_score = _score_against_set(benefit_key)
    total = round(price_score * 0.5 + benefit_score * 0.5, 1)
    return total, "price 50% · benefit 50%"


def run(payload: dict) -> dict:
    category_name = payload.get("category_name") or "Unknown"
    category_slug = payload.get("category_slug") or ""
    schema = payload.get("attribute_schema") or []
    listings = payload.get("listings") or []
    if len(listings) < 1:
        return {"text": "No listings to compare.\n", "winner": None, "runner_ups": [], "note": None}

    prices = [_num(l.get("price")) for l in listings]
    prices = [p for p in prices if p is not None]
    if not prices:
        prices = [0.0 for _ in listings]

    median_price = sorted(prices)[len(prices) // 2] if prices else 0.0

    scored: list[dict] = []
    for listing in listings:
        value_score, blend = compute_value_score(listing, category_slug, schema, prices, listings)

        # Plain, spoken-out-loud phrasing on purpose — this is the sentence
        # shown directly to the user as "Why", so it needs to read like an
        # explanation a ten-year-old would follow, not a scoring log. The
        # actual math (weights, bonuses) still lives in the full text dump
        # below for anyone who wants the audit trail.
        reasons: list[str] = []
        special = False
        trend = _num(listing.get("trend_percent"))
        if trend is not None and trend <= TREND_DROP_THRESHOLD:
            special = True
            reasons.append(f"its price just dropped {abs(trend):.0f}%, so it's cheaper than it was recently")
        price = _num(listing.get("price"))
        if price is not None and median_price > 0 and price <= median_price * BELOW_MEDIAN_FACTOR:
            special = True
            reasons.append("it costs less than most of the other options here")

        freshness = listing.get("freshness_status") or "unverified"
        freshness_bonus = FRESHNESS_BONUS.get(freshness, 0)
        if freshness_bonus > 0:
            reasons.append("the price we have for it is fresh and up to date")
        elif freshness == "stale":
            reasons.append("the price we have for it is a bit old, so it may have changed")
        elif freshness == "unverified":
            reasons.append("we haven't double-checked this price recently")

        verified = bool(listing.get("provider_verified"))
        trust_bonus = TRUST_BONUS if verified else 0
        if not verified:
            reasons.append("its provider hasn't been verified yet, so it's worth double-checking")

        bonus = SPECIAL_BONUS if special else 0

        final = max(0.0, min(100.0, round(value_score + bonus + freshness_bonus + trust_bonus, 1)))

        scored.append(
            {
                "listing": listing,
                "value_score": value_score,
                "blend": blend,
                "final": final,
                "bonus": bonus,
                "freshness_bonus": freshness_bonus,
                "trust_bonus": trust_bonus,
                "reasons": reasons,
            }
        )

    scored.sort(key=lambda s: s["final"], reverse=True)

    lines: list[str] = []
    lines.append(BAR)
    lines.append("TRADITIONAL COMPARISON — rules-based engine, no AI")
    lines.append(f"Category: {category_name}")
    lines.append("Weights: transparent per-category value blend + special/freshness/trust bonuses")
    lines.append(BAR)

    for rank, entry in enumerate(scored, start=1):
        listing = entry["listing"]
        provider = listing.get("provider") or "Unknown provider"
        lines.append("")
        lines.append(f"{rank}. {listing.get('name') or 'Unnamed'}  ({provider})")
        price = _num(listing.get("price"))
        currency = listing.get("currency") or ""
        price_txt = f"{price:,.2f} {currency}".strip() if price is not None else "price —"
        lines.append(f"   Price: {price_txt}")
        for attr in schema:
            if attr.get("is_comparable") and attr.get("key") != "price" and attr.get("data_type") == "number":
                if _num(listing.get("attributes", {}).get(attr["key"])) is not None:
                    lines.append(f"   {attr.get('label') or attr['key']}: {fmt_attr(listing, attr['key'], schema)}")
        lines.append(f"   Value score: {entry['value_score']}/100  ({entry['blend']})")
        detail = f"   Adjustments: special +{entry['bonus']}" if entry["bonus"] else "   Adjustments: none"
        detail += f" · freshness {entry['freshness_bonus']:+d} · trust {entry['trust_bonus']:+d}"
        lines.append(detail)
        lines.append(f"   FINAL: {entry['final']}/100")
        if entry["reasons"]:
            lines.append("   Because: " + speak(entry["reasons"]))

    lines.append("")
    lines.append(BAR)
    winner = scored[0]
    winner_why = speak(winner["reasons"]) if winner["reasons"] else "it scored best overall once we weighed price and features fairly"
    lines.append(f"WINNER: {winner['listing'].get('name')} — {winner['final']}/100")
    lines.append(f"Why: {winner_why}.")

    runner_ups = [
        {"name": s["listing"].get("name") or "Unnamed", "score": s["final"]}
        for s in scored[1:3]
    ]
    if runner_ups:
        lines.append("")
        lines.append("Runner-up: " + " — ".join(f"{r['name']} ({r['score']}/100)" for r in runner_ups))

    note = "This is a deterministic rules-based comparison, not an AI recommendation."
    lines.append("")
    lines.append(f"Note: {note[0].lower()}{note[1:]}")

    return {
        "text": "\n".join(lines) + "\n",
        "winner": {
            "name": winner["listing"].get("name") or "Unnamed",
            "score": winner["final"],
            "why": winner_why,
        },
        "runner_ups": runner_ups,
        "note": note,
    }


if __name__ == "__main__":
    try:
        payload = json.load(sys.stdin)
    except json.JSONDecodeError as exc:
        print(json.dumps({"error": f"could not read JSON payload: {exc}"}), file=sys.stderr)
        sys.exit(1)
    print(json.dumps(run(payload)))

File: src/lib/traditionalCompare.ts
import type { AttributeSchemaFieldDTO } from "@/types/catalog";
import { isLowerBetterAttribute } from "@/lib/attributeDirection";

/**
 * "Traditional comparison" — a deterministic, rules-based engine, deliberately
 * not an AI/LLM. This replaces the former scripts/traditional_compare.py,
 * which (a) could never actually run in the deployed image — the production
 * Dockerfile's runner stage installs no Python interpreter and never copies
 * scripts/ into the image — and (b) never inverted direction for
 * weighted "lower is better" attributes (monthly_fee, transaction_fee), so it
 * scored a higher fee as better. Running in-process fixes the deployment gap,
 * and importing isLowerBetterAttribute here — the same function
 * src/lib/scoring.ts already uses — fixes the direction bug by giving both
 * engines one shared source of truth instead of a second, drifted copy.
 */

export type TraditionalListingInput = {
  id: string;
  name: string | null;
  provider: string | null;
  providerVerified: boolean;
  price: number | null;
  currency: string | null;
  attributes: Record<string, unknown>;
  freshnessStatus: string | null;
  trendPercent: number | null;
};

export type TraditionalComparisonResult = {
  text: string;
  winner: { name: string; score: number; why: string } | null;
  runnerUps: { name: string; score: number }[];
  note: string | null;
};

// Per-category weights, lifted from notebooks/recommendation_engine.py so this
// on-demand engine agrees with the offline recommender. Categories without an
// entry fall back to the legacy 50/50 price + first-numeric-benefit blend.
const CATEGORY_WEIGHTS: Record<string, Record<string, number>> = {
  "data-bundles": { price: 0.4, data_amount: 0.4, validity_days: 0.2 },
  "voice-sms-bundles": { price: 0.35, minutes: 0.35, sms_count: 0.2, validity_days: 0.1 },
  "savings-accounts": { price: 0.3, interest_rate: 0.4, monthly_fee: 0.3 },
  "current-accounts": { price: 0.4, transaction_fee: 0.3, branch_count: 0.3 },
  "motor-insurance": { price: 0.35, coverage_amount: 0.45, claim_turnaround_days: 0.2 },
  "life-insurance": { price: 0.3, coverage_amount: 0.4, payout_speed_days: 0.3 },
};

const FRESHNESS_BONUS: Record<string, number> = { fresh: 5, stale: -6, unverified: -12 };
const SPECIAL_BONUS = 20;
const TRUST_BONUS = 5;
const TREND_DROP_THRESHOLD = -5;
const BELOW_MEDIAN_FACTOR = 0.9;

const BAR = "=".repeat(60);

/** Joins reason phrases into one plain sentence: "A", "A and B", or "A, B, and C". */
function speak(reasons: string[]): string {
  if (reasons.length === 0) return "";
  if (reasons.length === 1) return reasons[0];
  if (reasons.length === 2) return `${reasons[0]} and ${reasons[1]}`;
  return `${reasons.slice(0, -1).join(", ")}, and ${reasons[reasons.length - 1]}`;
}

function num(value: unknown): number | null {
  if (typeof value === "number") return Number.isFinite(value) ? value : null;
  if (typeof value === "string" && value.trim() !== "") {
    const n = Number(value);
    return Number.isFinite(n) ? n : null;
  }
  return null;
}

function normalize(values: number[], invert: boolean): number[] {
  const lo = Math.min(...values);
  const hi = Math.max(...values);
  if (lo === hi) return values.map(() => 100);
  return values.map((v) => {
    const t = (v - lo) / (hi - lo);
    const val = (invert ? 1 - t : t) * 100;
    return Math.round(val * 10) / 10;
  });
}

function firstNumericBenefitKey(schema: AttributeSchemaFieldDTO[]): string | null {
  const numeric = schema
    .filter((a) => a.dataType === "number" && a.isComparable && a.key !== "price")
    .sort((a, b) => a.sortOrder - b.sortOrder);
  return numeric[0]?.key ?? null;
}

function fmtAttr(listing: TraditionalListingInput, key: string, schema: AttributeSchemaFieldDTO[]): string {
  const value = listing.attributes[key];
  if (value === undefined || value === null) return "—";
  const unit = schema.find((a) => a.key === key)?.unit;
  return unit ? `${value} ${unit}`.trim() : `${value}`;
}

function computeValueScore(
  listing: TraditionalListingInput,
  categorySlug: string,
  schema: AttributeSchemaFieldDTO[],
  prices: number[],
  comparedSet: TraditionalListingInput[],
): { score: number; blend: string } {
  const weights = CATEGORY_WEIGHTS[categorySlug];
  const attrs = listing.attributes;
  const price = num(listing.price);
  if (price === null) return { score: 0, blend: "no price" };

  const priceScores = normalize(prices, true);
  const priceIdx = prices.indexOf(price);
  const priceScore = priceIdx !== -1 ? priceScores[priceIdx] : 100;

  const column = (key: string): number[] =>
    comparedSet.map((l) => num(l.attributes[key])).filter((v): v is number => v !== null);

  // Direction-aware, unlike the Python original: inverts the normalization
  // for any key attributeDirection.ts flags as "lower is better" (fees,
  // waiting periods, ...) instead of always treating a higher raw value as
  // the better one.
  const scoreAgainstSet = (key: string): number => {
    const col = column(key);
    const raw = num(attrs[key]);
    if (raw === null || col.length < 2) return 100;
    const norm = normalize(col, isLowerBetterAttribute(key));
    const idx = col.indexOf(raw);
    return idx !== -1 ? norm[idx] : 100;
  };

  if (weights) {
    const parts: number[] = [];
    const breakdown: string[] = [];
    for (const [key, weight] of Object.entries(weights)) {
      if (key === "price") {
        parts.push(priceScore * weight);
        breakdown.push(`price ${Math.round(weight * 100)}%`);
        continue;
      }
      const raw = num(attrs[key]);
      if (raw === null) continue;
      parts.push(scoreAgainstSet(key) * weight);
      breakdown.push(`${key} ${Math.round(weight * 100)}%`);
    }
    if (parts.length > 0) {
      const score = Math.round(parts.reduce((a, b) => a + b, 0) * 10) / 10;
      return { score, blend: breakdown.join(" · ") };
    }
    return { score: priceScore, blend: "price only" };
  }

  const benefitKey = firstNumericBenefitKey(schema);
  if (benefitKey === null) return { score: priceScore, blend: "price only (no benefit attribute)" };
  const raw = num(attrs[benefitKey]);
  if (raw === null) return { score: priceScore, blend: "price only" };
  const benefitScore = scoreAgainstSet(benefitKey);
  const total = Math.round((priceScore * 0.5 + benefitScore * 0.5) * 10) / 10;
  return { score: total, blend: "price 50% · benefit 50%" };
}

export function computeTraditionalComparison(params: {
  categoryName: string;
  categorySlug: string;
  attributeSchema: AttributeSchemaFieldDTO[];
  listings: TraditionalListingInput[];
}): TraditionalComparisonResult {
  const { categoryName, categorySlug, attributeSchema: schema, listings } = params;
  if (listings.length < 1) {
    return { text: "No listings to compare.\n", winner: null, runnerUps: [], note: null };
  }

  let prices = listings.map((l) => num(l.price)).filter((p): p is number => p !== null);
  if (prices.length === 0) prices = listings.map(() => 0);
  const sortedPrices = [...prices].sort((a, b) => a - b);
  const medianPrice = sortedPrices[Math.floor(sortedPrices.length / 2)] ?? 0;

  type Scored = {
    listing: TraditionalListingInput;
    valueScore: number;
    blend: string;
    final: number;
    bonus: number;
    freshnessBonus: number;
    trustBonus: number;
    reasons: string[];
  };

  const scored: Scored[] = listings.map((listing) => {
    const { score: valueScore, blend } = computeValueScore(listing, categorySlug, schema, prices, listings);

    // Plain, spoken-out-loud phrasing on purpose — this is the sentence shown
    // directly to the user as "Why", so it needs to read like an explanation
    // a ten-year-old would follow, not a scoring log.
    const reasons: string[] = [];
    let special = false;
    const trend = num(listing.trendPercent);
    if (trend !== null && trend <= TREND_DROP_THRESHOLD) {
      special = true;
      reasons.push(`its price just dropped ${Math.round(Math.abs(trend))}%, so it's cheaper than it was recently`);
    }
    const price = num(listing.price);
    if (price !== null && medianPrice > 0 && price <= medianPrice * BELOW_MEDIAN_FACTOR) {
      special = true;
      reasons.push("it costs less than most of the other options here");
    }

    const freshness = listing.freshnessStatus ?? "unverified";
    const freshnessBonus = FRESHNESS_BONUS[freshness] ?? 0;
    if (freshnessBonus > 0) {
      reasons.push("the price we have for it is fresh and up to date");
    } else if (freshness === "stale") {
      reasons.push("the price we have for it is a bit old, so it may have changed");
    } else if (freshness === "unverified") {
      reasons.push("we haven't double-checked this price recently");
    }

    const verified = Boolean(listing.providerVerified);
    const trustBonus = verified ? TRUST_BONUS : 0;
    if (!verified) {
      reasons.push("its provider hasn't been verified yet, so it's worth double-checking");
    }

    const bonus = special ? SPECIAL_BONUS : 0;
    const final = Math.max(
      0,
      Math.min(100, Math.round((valueScore + bonus + freshnessBonus + trustBonus) * 10) / 10),
    );

    return { listing, valueScore, blend, final, bonus, freshnessBonus, trustBonus, reasons };
  });

  scored.sort((a, b) => b.final - a.final);

  const lines: string[] = [];
  lines.push(BAR);
  lines.push("TRADITIONAL COMPARISON — rules-based engine, no AI");
  lines.push(`Category: ${categoryName}`);
  lines.push("Weights: transparent per-category value blend + special/freshness/trust bonuses");
  lines.push(BAR);

  scored.forEach((entry, i) => {
    const rank = i + 1;
    const { listing } = entry;
    const provider = listing.provider ?? "Unknown provider";
    lines.push("");
    lines.push(`${rank}. ${listing.name ?? "Unnamed"}  (${provider})`);
    const price = num(listing.price);
    const currency = listing.currency ?? "";
    const priceTxt =
      price !== null
        ? `${price.toLocaleString("en-US", { minimumFractionDigits: 2, maximumFractionDigits: 2 })} ${currency}`.trim()
        : "price —";
    lines.push(`   Price: ${priceTxt}`);
    for (const attr of schema) {
      if (attr.isComparable && attr.key !== "price" && attr.dataType === "number") {
        if (num(listing.attributes[attr.key]) !== null) {
          lines.push(`   ${attr.label || attr.key}: ${fmtAttr(listing, attr.key, schema)}`);
        }
      }
    }
    lines.push(`   Value score: ${entry.valueScore}/100  (${entry.blend})`);
    let detail = entry.bonus ? `   Adjustments: special +${entry.bonus}` : "   Adjustments: none";
    detail += ` · freshness ${entry.freshnessBonus >= 0 ? "+" : ""}${entry.freshnessBonus} · trust ${entry.trustBonus >= 0 ? "+" : ""}${entry.trustBonus}`;
    lines.push(detail);
    lines.push(`   FINAL: ${entry.final}/100`);
    if (entry.reasons.length > 0) {
      lines.push("   Because: " + speak(entry.reasons));
    }
  });

  lines.push("");
  lines.push(BAR);
  const winnerEntry = scored[0];
  const winnerWhy =
    winnerEntry.reasons.length > 0
      ? speak(winnerEntry.reasons)
      : "it scored best overall once we weighed price and features fairly";
  lines.push(`WINNER: ${winnerEntry.listing.name} — ${winnerEntry.final}/100`);
  lines.push(`Why: ${winnerWhy}.`);

  const runnerUps = scored.slice(1, 3).map((s) => ({ name: s.listing.name ?? "Unnamed", score: s.final }));
  if (runnerUps.length > 0) {
    lines.push("");
    lines.push("Runner-up: " + runnerUps.map((r) => `${r.name} (${r.score}/100)`).join(" — "));
  }

  const note = "This is a deterministic rules-based comparison, not an AI recommendation.";
  lines.push("");
  lines.push(`Note: ${note[0].toLowerCase()}${note.slice(1)}`);

  return {
    text: lines.join("\n") + "\n",
    winner: { name: winnerEntry.listing.name ?? "Unnamed", score: winnerEntry.final, why: winnerWhy },
    runnerUps,
    note,
  };
}
